writers/batchwriter/batchwriter.go

```diff
@@ -122,39 +122,45 @@ func (w *BatchWriter) Close(context.Context) error {
 }
 
 func (w *BatchWriter) worker(ctx context.Context, tableName string, ch <-chan *message.WriteInsert, flush <-chan chan bool) {
-	sizeBytes := int64(0)
-	resources := make([]*message.WriteInsert, 0, w.batchSize)
+	bytes, rows := writers.NewCapped(w.batchSizeBytes), writers.NewCapped(w.batchSize)
+	resources := make([]*message.WriteInsert, 0, w.batchSize) // at least we have 1 row per record
 	ticker := writers.NewTicker(w.batchTimeout)
 	defer ticker.Stop()
+
+	send := func() {
+		w.flushTable(ctx, tableName, resources)
+		clear(resources)
+		resources = resources[:0]
+		bytes.Reset()
+		rows.Reset()
+	}
 	for {
 		select {
 		case r, ok := <-ch:
 			if !ok {
-				if len(resources) > 0 {
+				if rows.Current() > 0 {
 					w.flushTable(ctx, tableName, resources)
 				}
 				return
 			}
 
-			if (w.batchSize > 0 && len(resources) >= w.batchSize) || (w.batchSizeBytes > 0 && sizeBytes+util.TotalRecordSize(r.Record) >= int64(w.batchSizeBytes)) {
-				w.flushTable(ctx, tableName, resources)
+			dataBytes := util.TotalRecordSize(r.Record)
+			if rows.OverflownBy(r.Record.NumRows()) || bytes.OverflownBy(dataBytes) {
```

Comment:
btw, we had inconsintency here as we checked for bytes overflow in advance, but not for the rows overflow

Reply:
> btw, we had inconsintency here as we checked for bytes overflow in advance, but not for the rows overflow

Not sure I understand this comment, we did `len(resources) >= w.batchSize)` then we appended so we checked the overflow for rows before appending

Reply:
we checked if the rows were already at the level (but yeah, it worked for the single row records & not for the multirow here).